torch/distributed/_tensor/debug/comm_mode.py

```diff
@@ -0,0 +1,61 @@
+from collections import defaultdict
+from typing import Any, Dict
+
+import torch
+from torch.utils._python_dispatch import TorchDispatchMode
+
+
+funcol = torch.ops.c10d_functional
+
+
+class CommDebugMode(TorchDispatchMode):
+    """
+    ``CommDebugMode`` is a context manager that counts the number of
+    communications within its context. It does this using a ``TorchDispatchMode``.
```

Comment:
My primary concern on this PR is summed up by "this docstring is misleading, but I'd rather have the Mode behave like the docstring rather than fix the docstring."  --  it only logs the 3 types of registered functional collectives, not all communications ops.  I can see users being confused why their CommDebugMode didn't capture some collectives when they run it with FSDP for example.

Any thoughts about expanding this to log c10d ops too?  would not be as easy..

Alternatively, make this tool more explicitly focused on DTensor comms and then promise it'll capture all DTensor comms?

Reply:
i guess a possible proposal is something like this

- build a func in c++ that marks 'collective called' and hook it up to all the c10d collectives
- bind the 'collective called' to python to also instrument funcol

And have 2 UIs for getting at the data (separately)
- add a 'TORCH_LOGS' mode that just log-prints every collective call, but with no user code edit
- build a 'counter obj' in c++ and bind it out to python, so you can turn on a contextmgr and have it record collectives only during the active range, then progrmatically query info about the collectives (possibly including c++ and py stacktraces of where they got called)

Reply:
Yeah I was thinking about covering the c10d ops later so that FSDP could also works under this, and also a bulk of missing features currently i.e. identifying module boundaries and print out module hierarchy, etc. let me try out c10d ops to see if is easy to add, if not, I can adjust the docstr to mention that it's currently only supports functional collectives or DTensor comms, where I can add all functional collectives to it. Then in follow up PRs figuring out how the c10d ops should work

Reply:
updated the comment to reflect that it only supports functional collectives atm, added a remaining functional collectives.